pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -165,46 +190,305 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 }
 
+func (m *topologyPairsMaps) clone() *topologyPairsMaps {
+	copy := newTopologyPairsMaps()
+	copy.appendMaps(m)
+	return copy
+}
+
+func (podSpreadMap *topologyPairsPodSpreadMap) addPod(addedPod, metapod *v1.Pod, node *v1.Node) error {
+	constraints := getHardTopologySpreadConstraints(metapod)
+	match, err := podMatchesAllSpreadConstraints(addedPod, metapod.Namespace, constraints)
+	if err != nil {
+		return err
+	}
+	if !match || !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+		return nil
+	}
+
+	// records which topology key(s) needs to be updated
+	minMatchNeedsUpdate := make(map[string]struct{})
+	for _, constraint := range constraints {
+		pair := topologyPair{
+			key:   constraint.TopologyKey,
+			value: node.Labels[constraint.TopologyKey],
+		}
+		// it means current node is one of the critical paths of minMatchMap[TopologyKey]
+		if int32(len(podSpreadMap.topologyPairToPods[pair])) == podSpreadMap.minMatchMap[pair.key] {
+			minMatchNeedsUpdate[pair.key] = struct{}{}
+		}
+		podSpreadMap.addTopologyPair(pair, addedPod)
+	}
+	// no need to addHeadlessTopologyPair b/c if a headless pair must be present,
+	// it should have already been created earlier in removePod() phase
+
+	// In most cases, min match map doesn't need to be updated.
+	// But it's required to be updated when current node is the ONLY critical path which impacts
+	// the min match. With that said, in this case min match needs to be updated to min match + 1
+	if len(minMatchNeedsUpdate) != 0 {
+		// TODO(Huang-Wei): performance can be optimized.
+		// A possible solution is to record number of critical paths which co-impact the min match.
+		tempMinMatchMap := make(map[string]int32, len(minMatchNeedsUpdate))
+		for key := range minMatchNeedsUpdate {
+			tempMinMatchMap[key] = MaxInt32
+		}
+		for pair, podSet := range podSpreadMap.topologyPairToPods {
+			if _, ok := minMatchNeedsUpdate[pair.key]; !ok {
+				continue
+			}
+			if l := int32(len(podSet)); l < tempMinMatchMap[pair.key] {
+				tempMinMatchMap[pair.key] = l
+			}
+		}
+		for key, tempMin := range tempMinMatchMap {
+			if tempMin == podSpreadMap.minMatchMap[key]+1 {
+				podSpreadMap.minMatchMap[key] = tempMin
+			}
+		}
+	}
+	return nil
+}
+
+func (podSpreadMap *topologyPairsPodSpreadMap) removePod(deletedPod *v1.Pod) {
+	if podSpreadMap == nil || deletedPod == nil {
+		return
+	}
+
+	deletedPodFullName := schedutil.GetPodFullName(deletedPod)
+	pairSet, ok := podSpreadMap.podToTopologyPairs[deletedPodFullName]
+	if !ok {
+		return
+	}
+	topologyPairToPods := podSpreadMap.topologyPairToPods
+	for pair := range pairSet {
+		delete(topologyPairToPods[pair], deletedPod)
+		// if topologyPairToPods[pair] is empty after deletion
+		// don't clean it up as that topology counts as a match now
+	}
+	delete(podSpreadMap.podToTopologyPairs, deletedPodFullName)
+
+	// re-calculate minMatches
+	// removal of the deletedPod would probably genereate a smaller matching number
+	for pair := range pairSet {
```

Comment:
It seems this loop can be folded into the loop starting line 461

Reply:
Thanks for the catch. Updated.